src/lib/device-online-state.ts

```diff
@@ -0,0 +1,328 @@
+	}
+
+	private consume() {
+		// pull a message from the queue...
+		this.rsmq
+			.receiveMessageAsync({
+				qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+				vt: DeviceOnlineStateManager.RSMQ_READ_TIMEOUT, // prevent other consumers seeing the same message (if any) preventing multiple API agents from processing it...
+			})
+			.then(msg => {
+				if ('id' in msg) {
+					const { id, message } = msg;
+
+					return Bluebird.try(() => {
+						const { uuid, nextState } = JSON.parse(message) as {
+							uuid: string;
+							nextState: DeviceOnlineStates;
+						};
+
+						// raise and event for the state change...
+						switch (nextState) {
+							case DeviceOnlineStates.Timeout:
+								this.scheduleChangeOfStateForDevice(
+									uuid,
+									DeviceOnlineStates.Timeout,
+									DeviceOnlineStates.Offline,
+									API_HEARTBEAT_STATE_TIMEOUT_SECONDS, // put the device into a timeout state if it misses it's scheduled heartbeat window... then mark as offline
+								);
+								this.deviceTimeout(uuid);
+								break;
+							case DeviceOnlineStates.Offline:
+								this.deviceOffline(uuid);
+								break;
+							default:
+								captureException(
+									new Error(
+										`An unexpected value was encountered for the target device state: ${nextState}`,
+									),
+								);
+						}
+					}).finally(() =>
+						this.rsmq.deleteMessageAsync({
+							qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+							id,
+						}),
+					);
+				} else {
+					// no message! this is thrown to restart the consume process...
+					return Bluebird.delay(1000).thenThrow(
+						new NoMessagesToConsumeSignal(),
+					);
+				}
+			})
+			.catch(err => {
+				if (err instanceof NoMessagesToConsumeSignal) {
+					return;
+				}
+
+				captureException(
+					err,
+					'DeviceStateManager ==> Error when consuming queue',
+				);
+			})
+			.then(() => this.consume());
+	}
+
+	private scheduleChangeOfStateForDevice(
+		uuid: string,
+		currentState: DeviceOnlineStates,
+		nextState: DeviceOnlineStates,
+		delay: number, // in seconds
+	) {
+		// remove the old queued state...
+		return this.redis
+			.getAsync(`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`)
+			.then(value => {
+				const { id } = JSON.parse(value) as { id: string };
+
+				if (id) {
+					return this.rsmq.deleteMessageAsync({
+						qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+						id,
+					});
+				}
+
+				return 1;
+			})
+			.catch(noop) // ignore errors when deleting the old queued state and queue the new one...
+			.then(() =>
+				this.rsmq.sendMessageAsync({
+					qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+					message: JSON.stringify({
+						uuid,
+						nextState,
+					}),
+					delay,
+				}),
+			)
+			.then(id =>
+				this.redis
+					.setAsync(
+						`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`,
+						JSON.stringify({
+							id,
+							currentState,
+						}),
+					)
+					.then(() =>
+						this.redis.expireAsync(
```

Comment:
I think this can be combined into the `setAsync` via
```ts
				this.redis
					.setAsync(
						`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`,
						JSON.stringify({
							id,
							currentState,
						}),
						'EX',
						delay + 5
					)

Reply:
Added type definition to support this, and now using it.